src/Language/PureScript/TypeChecker/Kinds.hs

```diff
@@ -951,7 +955,8 @@ kindsOfAll moduleName syns dats clss = withFreshSubstitution $ do
           (args', supers', decls', generalizeUnknownsWithVars unkBinders clsKind)
     datResultsWithKinds <- for datResultsWithUnks $ \(((datName, datKind), ctors), _) -> do
       let tyUnks = snd . fromJust $ lookup (mkQualified datName moduleName) tySubs
-          ctors' = fmap (fmap (generalizeUnknowns tyUnks . replaceTypeCtors)) ctors
+          replaceTypeCtorsAndGeneralizeUnknowns = generalizeUnknowns tyUnks . replaceTypeCtors
+          ctors' = fmap (mapDataCtorFields (fmap (fmap replaceTypeCtorsAndGeneralizeUnknowns)) *** replaceTypeCtorsAndGeneralizeUnknowns) ctors
```

Comment:
It's not clear to me that you want to generalize unknowns in the data constructor spine, as this will introduce `forall`s inside the data constructor spine for any field with an unknown. These unknowns should be quantified as part of the data declaration, not in each field. That is
```
data Foo f a = Foo (f a)
```
This should be quantified as
```
data Foo :: forall k. (k -> Type) -> k -> Type
data Foo f a = Foo (f @k a)
```
Not as
```
data Foo f a = Foo (forall k. f @k a)
```

Reply:
Oh that’s why I had those unwanted foralls! I’ve replaced unknowns without generalizing them.